dot/sync/chain_sync_test.go

```diff
@@ -1271,13 +1277,82 @@ func Test_chainSync_setBlockAnnounce(t *testing.T) {
 				}
 			},
 		},
+		"err_when_calling_has_header": {
+			wantErr: errors.New("checking header exists"),
+			args: args{
+				header: blockAnnounceHeader,
+			},
+			chainSyncBuilder: func(ctrl *gomock.Controller) chainSync {
+				mockBlockState := NewMockBlockState(ctrl)
+				mockBlockState.EXPECT().
+					HasHeader(common.MustHexToHash(
+						"0x05bdcc454f60a08d427d05e7f19f240fdc391f570ab76fcb96ecca0b5823d3bf")).
+					Return(false, errors.New("checking header exists"))
+				mockDisjointBlockSet := NewMockDisjointBlockSet(ctrl)
+				return chainSync{
+					blockState:    mockBlockState,
+					pendingBlocks: mockDisjointBlockSet,
+				}
+			},
+		},
+		"adding_block_header_to_pending_blocks": {
+			args: args{
+				header: blockAnnounceHeader,
+			},
+			chainSyncBuilder: func(ctrl *gomock.Controller) chainSync {
+				argumentHeaderHash := common.MustHexToHash(
+					"0x05bdcc454f60a08d427d05e7f19f240fdc391f570ab76fcb96ecca0b5823d3bf")
+
+				mockBlockState := NewMockBlockState(ctrl)
+				mockBlockState.EXPECT().
+					HasHeader(argumentHeaderHash).
+					Return(false, nil)
+
+				mockBlockState.EXPECT().
+					BestBlockHeader().
+					Return(&types.Header{Number: 1}, nil)
+
+				mockDisjointBlockSet := NewMockDisjointBlockSet(ctrl)
+				mockDisjointBlockSet.EXPECT().
+					addHeader(blockAnnounceHeader).
+					Return(nil)
+
+				mockDisjointBlockSet.EXPECT().
+					addHashAndNumber(argumentHeaderHash, uint(2)).
+					Return(nil)
+
+				return chainSync{
+					blockState:    mockBlockState,
+					pendingBlocks: mockDisjointBlockSet,
+					peerState:     make(map[peer.ID]*peerState),
+					// creating an buffered channel for this specific test
+					// since it will put a work on the queue and an unbufered channel
+					// will hang until we read on this channel and the goal is to
+					// put the work on the channel and don't block
+					workQueue: make(chan *peerState, 1),
+				}
+			},
+		},
 	}
 	for name, tt := range tests {
 		t.Run(name, func(t *testing.T) {
+			t.Parallel()
 			ctrl := gomock.NewController(t)
 			sync := tt.chainSyncBuilder(ctrl)
 			err := sync.setBlockAnnounce(tt.args.from, tt.args.header)
-			assert.ErrorIs(t, err, tt.wantErr)
+			if tt.wantErr != nil {
+				assert.EqualError(t, err, tt.wantErr.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+
+			if sync.workQueue != nil {
+				assert.LessOrEqual(t, len(sync.workQueue), 1)
+				if len(sync.workQueue) > 0 {
```

Comment:
meganit you could have this as a for loop instead (change `if` to `for` and remove `close` of channel), in case there is more than 1 element in the channel

Reply:
This channel cannot be closed easily by the writer unless we do some refactoring in this code, this channel is created and read based on a timer, so if I put a `for work := range sync.workQueue` then it will block until we close the channel which never happens, I will keep it as it is but removing the `close`, something like:

```go
// if workQueue is not nil then I expect one element in the channel
if sync.workQueue != nil {
	assert.Equal(t, len(sync.workQueue), 1)
}
```